lib/quicly.c

```diff
@@ -1253,6 +1269,13 @@ static int record_receipt(quicly_conn_t *conn, struct st_quicly_pn_space_t *spac
         conn->egress.send_ack_at = conn->stash.now + QUICLY_DELAYED_ACK_TIMEOUT;
     }
 
+    if (is_ack_only)
+        ++conn->super.stats.num_packets.ack_only_received;
+    if (epoch == QUICLY_EPOCH_INITIAL || epoch == QUICLY_EPOCH_HANDSHAKE)
+        ++conn->super.stats.num_packets.handshake_received;
+    if (epoch == QUICLY_EPOCH_0RTT)
+        ++conn->super.stats.num_packets.zerortt_received;
```

Comment:
I'm not sure if this is the correct location for updating the metrics.

We update `num_packets.received` in `quicly_accept` and `quicly_receive`. Maybe we should change that to per-packet-type metric (i.e. `uint64_t [QUICLY_NUM_EPOCHS]`) and increment the counter of the corresponding epoch.